### ISSUE TITLE
incorrect resolution of subprocess.check_output() type when using **kwargs

### ISSUE THREAD

**Issue by cjp256** [microsoft/pyright#1849]
**Describe the bug**

pyright seems to incorrectly match on an overloaded signature for check_output() when using `**kwargs` to pass keyword arguments.

**To reproduce**

```
import subprocess
from typing import Dict

# Example 1: Pyright passes OK
x: bytes = subprocess.check_output(["echo", "-n", "hi"], text=False)
assert x == b'hi'

# Example 2: Pyright fails unexpectedly with
# error: Expression of type "str" cannot be assigned to declared type "bytes"
#    "str" is incompatible with "bytes" (reportGeneralTypeIssues)
kwargs: Dict = {}
y: bytes = subprocess.check_output(["echo", "-n", "hi"], text=False, **kwargs)
assert y == b'hi'
```

Playing around, it _appears_ to match on the first definition that has a required keyword-argument that is not actually present, such as:
https://github.com/microsoft/pyright/blob/master/packages/pyright-internal/typeshed-fallback/stdlib/subprocess.pyi#L432  where IIUC `encoding` is required.  If I remove that signature, it falls to the next one where `errors` is required, and so on.

**VS Code extension or command-line**
command-line, 1.1.138 (from git)

**Additional context**

Interestingly enough, if `mypy` doesn't have a type specified for `kwargs`, it will fail in a similar manner.  After defining `kwargs` as a Dict, mypy passes as expected.  However, as shown in the example, it does not correct it for pyright.

Originally reported by @cmatsuoka.

**Comment by erictraut** [comment]
After further inspection, I think this behavior is correct — or at least is the intended behavior in this case. Mypy has some odd rules for overload matching when "Any" values are concerned, and pyright intentionally does not duplicate the exact behavior. The overload matching behavior is not dictated in PEP 484, so there is no "standard" here.

I did find a [related bug](https://github.com/microsoft/pyright/issues/1861) in pyright when I was investigating this bug report. That other bug did impact the overload matching logic in this case, and it will be fixed in the next release.

The `check_output` function is defined in the `subprocess.pyi` type stub, and it consists of 6 overloads. Each overload signature is very complex, but there are just a few parameters that matter for the purposes of this analysis, so I've stripped out the irrelevant details here.

```python
@overload # Overload #1
def check_output(
    args: _CMD,
    *,
    encoding: Optional[str] = ...,
    text: Literal[True],
) -> str: ...
@overload # Overload #2
def check_output(
    args: _CMD,
    *,
    encoding: str,
    text: Optional[bool] = ...,
) -> str: ...
@overload # Overload #3
def check_output(
    args: _CMD,
    *,
    encoding: Optional[str] = ...,
    text: Optional[bool] = ...,
) -> str: ...
@overload # Overload #4
def check_output(
    args: _CMD,
    *,
    encoding: Optional[str] = ...,
    text: Optional[bool] = ...,
) -> str: ...
@overload # Overload #5
def check_output(
    args: _CMD,
    *,
    encoding: None = ...,
    text: Literal[None, False] = ...,
) -> bytes: ...
@overload # Overload #6
def check_output(
    args: _CMD,
    *,
    encoding: Optional[str] = ...,
    text: Optional[bool] = ...,
) -> Any: ...
```

Overloads are matched in order, so let's look at what's happening when you invoke the function with the arguments `check_output(["echo", "-n", "hi"], text=False, **kwargs)`.

Overload 1 does not match because `text=False` is incompatible with `Literal[True]`.

Overload 2 is where things get interesting. In this case, the `encoding` keyword parameter has no default value, so it must be provided a value for the overload to apply. The `**kwargs` argument may or may not be compatible with `encoding` depending on whether the values in the `kwargs` dictionary are compatible with `str`. So if `kwargs` is declared as `Dict[str, int]`, it will not match. But if `kwargs` is declared as `Dict[str, str]`, it will match. But what if `kwargs` is declared as `Dict[str, Any]` or `Dict[Unknown, Unknown]`? Here is where pyright and mypy differ in a subtle way. Pyright applies normal type assignment rules and determines that this overload does apply. Mypy, on the other hand, continues to search in this case. Mypy ends up determining that the return type for the call is `Any`.

I understand why the mypy authors chose this behavior, but I'm not convinced that it's the right choice, especially given that pyright has a more sophisticated way of tracking implicit `Any` values (which appear in error messages as `Unknown`). So at this time, I'm not inclined to adopt mypy's exact behavior here. This is the first time I've received a bug report that highlights this subtle difference. If we receive more of these reports over time, I might reconsider.

The ordering of the `check_output` overload looks wrong to me. I think that "Overload #5" should be moved much earlier, perhaps after "Overload #1". But I admit that I don't have expertise with this API. If you agree with my assessment, I encourage you to consider filing a bug and/or submitting a PR in the typeshed repo.

In the meantime, you can work around this issue in a couple of different ways:
1. Define `kwargs` with a type that will not match the type of the `encoding` parameter, such as `kwargs: Dict[str, bool] = {}`.
2. Use a call to `cast` to cast the output type from `str` to `bytes`.
3. Suppress the diagnostic using a `# type: ignore` comment.
